proposals/0000-numeric-underscores.rst

```diff
@@ -0,0 +1,173 @@
+.. proposal-number:: Leave blank. This will be filled in when the proposal is
+                     accepted.
+
+.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+
+.. highlight:: haskell
+
+This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/76>`_.
+
+.. contents::
+
+Underscores in Numeric Literals
+===============================
+
+GHC supports various numeric literals such as decimal, octal, hexadecimal, binary, and floating point numbers.
+However, large numeric literals are hard to read.
+This proposal improves the readability, quality, expressiveness of numeric literals.
+
+Motivation
+------------
+Large numeric literals are hard to read.
+This easily causes a bug.
+For example, the following numbers are confusing:
+
+.. code-block:: none
+
+    x = 10000000            -- Is this one million or ten million?
+    y = 0x3ffffff           -- Is this 4M or 64M?
+    z = 0b000100000000      -- Which position is a 1-bit digit?
+
+Readability can be improved by separating numbers with underscores (``_``).
+This feature improves readability, quality and expressiveness as follows:
+
+.. code-block:: none
+
+    x = 10_000_000          -- ten million
+    y = 0x3ff_ffff          -- 64M
+    z = 0b0001_0000_0000    -- 8th bit
+
+
+Similar feature is introduced in some languages such as `Verilog-HDL <https://inst.eecs.berkeley.edu/~cs150/fa06/Labs/verilog-ieee.pdf#page=20>`_ and `Rust <https://doc.rust-lang.org/reference/tokens.html#number-literals>`_ .
+
+Proposed Change Specification
+-----------------------------
+
+This proposal
+~~~~~~~~~~~~~
+I propose an extension to the existing syntax of numeric literals.
+
+Current syntax:
+
+.. code-block:: none
+
+    decimal     →  digit{digit}
+    octal       →  octit{octit}
+    hexadecimal →  hexit{hexit}
+    binary      →  binit{binit}
+
+New syntax (this proposal):
+
+.. code-block:: none
+
+    decimal     →  digit{[_ | digit]}
```

Comment:
This would allow trailing underscore, i.e. `1000_1000_`, do we want such thing? If we do, then there should be an example in proposal.

Reply:
Thank you.
For float only, I assumed the trailing underscore.

For example, the following description is possible.
  gravity = 6.674_08_e−11